### DISCUSSION TITLE
Integrate cython modules in pyvista

### DISCUSSION THREAD

**Discussion by Louis-Pujol** [pyvista/pyvista#5393] [Ideas]
Hi !

I was wondering about the opinion of pyvista developers about the integration of cython modules. It seems that there is no cython in pyvista/pyvista repository but on other repositories of the organization (https://github.com/pyvista/fast-simplification https://github.com/pyvista/pymeshfix). Is it a rule or would you also consider compiling cython modules in the main repository ?

My question is motivated by the discussion in #5376. I did a quick [POC](https://github.com/Louis-Pujol/cython_ncells_from_cells) that shows that the running time of the function [`n_cells_from_cells`](https://github.com/pyvista/pyvista/blob/main/pyvista/core/utilities/cells.py) can be greatly improved (~5x faster for large cell connectivity arrays) by writing it in cython.

I would like to integrate it, but I want feedback from the community about the way it should be done. And I think this discussion can serve as a basis to maybe write a paragraph in Contributing or Development Pratices to clarify the situation (and maybe address the question of the use of related tools, as JIT-compilers).

For the ones not familiar with cython, it is a language very close to python that can be translated into C and compiled to accelerate some parts of the code. The typical use case is when we need to run a loop on a large array and when there is no alternative with vectorized operations through Numpy. As an example, ts is used in Scikit-Learn  (see the documentation : [How to optimize for speed](https://scikit-learn.org/stable/developers/performance.html) and [Cython best practices, convention and knowledge](https://scikit-learn.org/stable/developers/cython.html))

Thanks,

**Comment by banesullivan** [comment]
> opinion of pyvista developers about the integration of cython modules

I have a hard stance that the main `pyvista` package must remain pure Python. Full stop.

The intent of `pyvista` is to provide a thin, accessible layer on top of VTK. Any Cython, C/C++, or code requiring a more involved build process or dependencies other than what we have already established should be contributed directly upstream to VTK (C++) or shipped as a standalone package depending on and extending PyVista. 

Another (somewhat selfish) reason for this stance is to prevent the added maintenance burden of Cython and having an actual build process. PyVista is maintained by volunteers at this time (🙋🏻‍♂️), and we do not have the bandwidth to maintain additional build complications for niche features.

>  It seems that there is no cython in pyvista/pyvista repository but on other repositories of the organization 

This is by design :) With Cython comes a plethora of build, deployment, and installation complications. Since PyVista is intended to be incredibly easy to use and the first stop for people new to the VTK world, we want to keep PyVista as lean and easy to install as possible. Any additional complications should be new packages in their own right. With precedence being set by PyVistaQt (adds Qt dependencies), PyMeshFix (Cython + incompatible license), PyACVD (Cython), etc

> My question is motivated by the discussion in https://github.com/pyvista/pyvista/pull/5376. I did a quick [POC](https://github.com/Louis-Pujol/cython_ncells_from_cells) that shows that the running time of the function [n_cells_from_cells](https://github.com/pyvista/pyvista/blob/main/pyvista/core/utilities/cells.py) can be greatly improved (~5x faster for large cell connectivity arrays) by writing it in cython.

This is excellent!! Nice work! My gut reaction is to wonder if you should contribute this functionality upstream in VTK as C++ code so that PyVista can then wrap that functionality. 

> serve as a basis to maybe write a paragraph in Contributing or Development Pratices

This is a great point and seems to be coming up a good bit recently around other dependency-complicating features: Cython, Qt, etc. I'll make a mental note to do this soon here at least.

----- 

> I want feedback from the community about the way it should be done

One approach I recommend is to check out PyVista's API for extending or overriding classes: https://docs.pyvista.org/version/stable/examples/99-advanced/extending-pyvista#extending-pyvista-example

This should enable you to subclass a PyVista class to implement the functionality you have so massively improved. @MatthewFlamm might be the best person to talk to about implementing this, as they designed the interface in #1584.

This begs the question in my mind of whether we should create a new `pyvista-optimized` package to house all kinds of optimizations like this, then have a simple `pip install pyvista-optimized` register an entry point that will automatically extend PyVista's classes with overridden, optimized methods like this but its more "opt-in" rather than imposing a complicated build/installation on all of our users and the added maintenance burden for those of us doing this strictly on a volunteer basis at this time. I'd be happy to help set up some general repository structure, hosting under the PyVista organization, and deployment if you're on board -- let me know! But I won't be able to provide long-term maintenance myself.

There's another, more nuanced and perhaps complicated approach via VTK directly that I outlined in: https://github.com/pyvista/pyvista/issues/3809#issue-1527631748 but I cannot recommend this approach as it may result in complications with other code in the VTK ecosystem (and PyVista if we decide to leverage this interface).

**Comment by Louis-Pujol** [comment]
Hi @banesullivan, thanks a lot for your detailed answer.

In my case, I tried to accelerate a function that is called during the initialization of PolyData, before VTK to compute the number of cells before the actual initialization of vtkCellArray.  Indeed, this part of the code was already marked as a candidate for Cython :)

https://github.com/pyvista/pyvista/blob/b594aac86953875b90b91200a586679d0b15468b/pyvista/core/cell.py#L630-L637

As you suggested I'll have a look at VTK implementation to see if any workaround exists or if something can be changed directly in VTK.

About the Pyvista interface: alternative ways to initialize PolyData with large connectivity arrays exists (and I think cover all the situations), so a solution with a warning + explanations in the documentation to handle large datasets would be sufficient to avoid the bottleneck here.

However, as you said, if such a `pyvista-optimized` package exists it can be useful to address other bottlenecks. I don't have an overview on the current performance bottlenecks so if you have something in mind, let me know, my current position would be:
- if no other parts of the code need Cython, it is perhaps to heavy to create and maintain a `pyvista-optimized` as workarounds already exists
- if it appears that there are other performance improvements that can be obtain this way, I can help maintaining `pyvista-optimized`

**Reply by banesullivan** [reply]
There are quite a few places in PyVista that could be improved by Cython but we haven't been keeping track of these in any formal way. One that comes to mind is:

https://github.com/pyvista/pyvista/blob/1115340a6995c5d205c398203cf0fd6c62a0e4df/pyvista/core/utilities/arrays.py#L574-L586

Perhaps we let this discussion sit for a bit and see if others see places to optimize PyVista and are able to commit to the implementation and maintenance. If we get 3 solid used cases (yours being 1) then I think it'd be a great opportunity to make a new `pyvista-optimized` package. 

I've gone ahead and made a new repository here: https://github.com/pyvista/pyvista-optimized. PRs always welcome and when it comes time to ship a Python package to PyPI I can set all of that up

**Comment by adeak** [comment]
For reference a few times when similar things came up with me involved:
  * https://github.com/pyvista/pyvista/issues/2330#issuecomment-1066168815 and nearby discussion
  * a longer acceleration-related discussion at https://github.com/pyvista/pyvista/pull/1213, with @akaszynski's bottom line at https://github.com/pyvista/pyvista/pull/1213#issuecomment-792042348

Keeping cython away from the main repo is still the best idea in my opinion, because cython code drastically reduces accessibility of the codebase.

**Reply by Louis-Pujol** [reply]
Thanks @adeak  for these references. The first one seems very close to [`n_cells_from_cells`](https://github.com/pyvista/pyvista/blob/main/pyvista/core/utilities/cells.py), except that we need to be able to output list of irregular faces. I never try to construct irregular arrays in cython, I think this can be done using vectors from the C++ standard library, but I wonder how to efficiently pass it to python afterwards.

Agree with @akaszynski that cython is the best solution for accelerating those kind of routines. I've also tried a bit numba, but for me cython is a best solution both in term of performance and in term of user-friendliness:
- Numba JIT-compiled functions are fast but compilation time is not. It is a good solution for functions that are run several times in the same session. If the compiled function is just executed one or twice, it could be slower than pure python. It could be optimized as Numba as a cache mechanism, but it seemed to me that there was still an overhead for loading the cache
- As cython relies on converting python to C or C++ and compile, once the compilation is done the fast function is accessible with no overhead
- Numba is required as a dependency for the user (and if I remember correctly, it has at some point very unpleasant compatibility issues with numpy)
- With cython, the burden is 100% on the developer side, cython is a dependence for the build process. If building goes right, the compiled modules are just shipped with the rest of the package and the user does not need to install cython to run the compiled function

For the clarity of the codebase, I agree that separating cython from the rest is a good idea.

**Comment by Louis-Pujol** [comment]
I tried to optimize the extraction of a list of irregular faces from a connectivity array https://github.com/pyvista/pyvista/issues/2330#issuecomment-1066168815 in cython.

It appeared to be very time-consuming to construct the structure and in some way memory-inefficient too because the indices of faces are duplicated from connectivity array to the new structure (list of lists or lists of numpy arrays). What can be obtain for free in `n_cells_from_cells` is the successive offsets (length of each face).

A workaround I found to solve this issue was to write a `JaggedArray` class. It is actually a wrapper around a connectivity array with the sequence of offsets. This class has `__getitem__` and `__len__` methods. The `__getitem__[i]` method simply return a view on the `i`th cell of the connectivity array (easy to compute the bounds thanks to the offsets), there is no data duplication. It is compatible with [`from_irregular_faces`](https://github.com/pyvista/pyvista/pull/5375), as shown by this [test](https://github.com/Louis-Pujol/cython_ncells_from_cells/blob/main/tests/test_jagged_array.py). The combination of (`n_cells_from_cells` + store the offsets in cython) / `JaggedArray` representation seems to me a good way to solve efficiently the issue raised in https://github.com/pyvista/pyvista/issues/2330#issuecomment-1066168815.

I also think that `JaggedArray`  could be an interesting tool to have in pyvista to ease transfers between 1D connectivity arrays and irregular faces lists of arrays in the continuity of what was done in https://github.com/pyvista/pyvista/pull/4637 and https://github.com/pyvista/pyvista/pull/5375. @adeak @darikg @tkoyama010 @MatthewFlamm @banesullivan as you were participating in these discussions, what are your opinions about  that ?

I'll try to find some time the next days to start a PR in `pyvista-optimize` for the integration of `n_cells_from_cells`.

**Reply by adeak** [reply]
This is mostly a drive-by comment because unfortunately I'm chronically short on time to look into things :( Your `JaggedArray` makes me wonder why you shouldn't just use a `dtype=object` NumPy array for this. Did you try that? It should also only contain references, complete with the full ndarray API for indexing and whatnot. It's a lot slower than a proper numerical dtype ndarray, but I'd expect something hand-written in Python to be similar or even worse.

Of course you'd then have to compute all the offset stuff upfront, because the data in the array would be static (in the sense that you couldn't have offset setters and whatnot).

**Reply by Louis-Pujol** [reply]
Thanks for the idea ! Didn't think about using numpy array with `dtype=object`. I did a quick test, assuming we have the connectivity array + the cell lengths + the offsets (everything is available in `CellArray`). Initialization of irregular_faces with `numpy` and `JaggedArray` look like:
```python
irregular_faces_np = np.array(
    [connectivity[offsets[i]+1:offsets[i] + cell_lengths[i]+1] for i in range(n_cells)],
    dtype=object,
)
irregular_faces_jagged = JaggedArray(connectivity, cell_lengths=cell_lengths)
```

With 10M cells of size 1-3, the running time of these lines is:
```
numpy (creation): 0.683 s
JaggedArray (creation): 0.015 s
```
Not surprising `JaggedArray`  is faster here. In order to create the numpy array I needed to run a loop over all cells position. `JaggedArray` is lazy, accesses to cells location are done only in `__getitem__`. Perhaps there is a smarter way to initialize the numpy array.

But when it comes from executing a function over all cells, `len` for example:
```python
# Numpy with for loop:
lengths = np.zeros(n_cells, dtype=int)
for i in range(n_cells):
    lengths[i] = len(irregular_faces_np[i])
end = time()

# Numpy with vectorized operation
vec_len = np.vectorize(len)
lengths = vec_len(irregular_faces_np)

# JaggedArray with for loop
lengths = np.zeros(n_cells, dtype=int)
for i in range(n_cells):
    lengths[i] = len(irregular_faces_jagged[i])
```

The running times looks as follows:
```
numpy (loop): 0.173 s
numpy (vectorized): 0.088 s
JaggedArray (loop): 0.820 s
```

So my conclusion with this quick test:

- `JaggedArray` is faster to create thanks to lazyness
- numpy array with `dtype=object` is slower to create (perhaps because my implementation is bad), but faster when we need to apply a function in parallel on every cells

I think that most of the time, we need to apply functions over all the cells and not only want a view on a small number of cells. Then I am most in favor of the numpy solution here, and as you mentioned, it gives access to the numpy API  (vectorize being an example of advantage here).

**Reply by adeak** [reply]
> `JaggedArray` is lazy, accesses to cells location are done only in `__getitem__`

In that case, yeah, there's no beating a lazy wrapper like that. But then there's the trade-off that using the same index twice reruns the same logic. If you're fine with a static array values could be cached perhaps.

> Perhaps there is a smarter way to initialize the numpy array.

Maybe not smarter but slightly more robust: initialize `np.empty(n_cells, dtype=object)` and then set each value in a loop. When the input to `np.array(..., dtype=object)` is actually compatible with a non-jagged array, numpy will sometimes do some dtype conversions rather than merely building an object-dtype array containing each element in the input list. But this might end up being even slower, not sure.

As for vectorized `len` calculation, I'm a bit surprised that the loop is that much slower than `vectorize`, since the latter is still pretty much a loop. Perhaps the issue is that iterating numpy arrays in Python is very slow. Could you please also time `lengths_lst = [len(arr) for arr in irregular_faces_np]` and also `np.array(<the same thing here>)` (assuming this indeed does the same thing)?

And as usual, caveat: your timings measure fast things, so there's a lot of error in there, especially from a single run. You might want to use a larger sample size, even if this is a quick timing comparison.

**Reply by Louis-Pujol** [reply]
Thanks for your comment; My bad for the worst quality of the comparison, what is more, I initialized `lengths` arrays with `dtype=int`, it surely caused extra computation and falsifies the results.

Here is a cleaner comparison, with mean and standard deviation of running times over 10 repetitions

I implemented the initialization with a numpy array the way you suggested me, and compare the application of `len` to each elements different ways :

- Initialize a `lengths` array with zeros and fill it using a `for` loop
- Define `lengths` with a list comprehension : `[len(arr) for arr in irregular_faces_np]`
- Define `lengths` with a list comprehension + convert it to `np.array`: `np.array([len(arr) for arr in irregular_faces_np])`

I did the same thing for `JaggedArray`. Here is the code and the output:

```python
import numpy as np
from time import time
from ncells_from_cells import JaggedArray

def generate_random_cells(n_cells: int) -> np.ndarray:
    """Generate a random array of cells."""
    cell_lengths = np.random.randint(1, 4, n_cells)
    cells = np.random.randint(0, 10, cell_lengths.sum() + n_cells, dtype=np.int64)
    offsets = np.concatenate([[0],np.cumsum(cell_lengths + 1)[: -1]])

    cells[offsets] = cell_lengths

    return cells, offsets, cell_lengths


time_numpy_creation = []
time_numpy_loop = []
time_numpy_loop_array = []
time_numpy_vectorized = []
time_jagged_creation = []
time_jagged_loop = []

n_cells = 1000000
reps = 10

for _ in range(reps):

    connectivity, offsets, cell_lengths = generate_random_cells(n_cells)

    start = time()
    # Start with an empty array of objects with the right length
    irregular_faces_np = np.empty(
        n_cells,
        dtype=object,
    )
    # Fill the array with the right faces
    for i in range(n_cells):
        irregular_faces_np[i] = connectivity[offsets[i]+1:offsets[i] + cell_lengths[i]+1]
    end = time()
    time_numpy_creation.append((end - start) / reps)

    start = time()
    lengths1 = [len(irregular_faces_np[i]) for i in range(n_cells)]
    end = time()
    time_numpy_loop.append((end - start) / reps)

    start = time()
    lengths2 = np.array([len(irregular_faces_np[i]) for i in range(n_cells)])
    end = time()
    time_numpy_loop_array.append((end - start) / reps)

    vec_len = np.vectorize(len)

    start = time()
    lengths = vec_len(irregular_faces_np)
    end = time()
    time_numpy_vectorized.append((end - start) / reps)

    start = time()
    irregular_faces_jagged = JaggedArray(connectivity, cell_lengths=cell_lengths)
    end = time()
    time_jagged_creation.append((end - start) / reps)


    start = time()
    lengths = [len(irregular_faces_jagged[i]) for i in range(n_cells)]
    end = time()
    time_jagged_loop.append((end - start) / reps)


print(f'numpy creation: {np.mean(time_numpy_creation):.4f} s +- {np.std(time_numpy_creation):.4f} s')
print(f'numpy loop: {np.mean(time_numpy_loop):.4f} s +- {np.std(time_numpy_loop):.4f} s')
print(f'numpy loop + array: {np.mean(time_numpy_loop_array):.4f} s +- {np.std(time_numpy_loop_array):.4f} s')
print(f'numpy vectorized: {np.mean(time_numpy_vectorized):.4f} s +- {np.std(time_numpy_vectorized):.4f} s')
print(f'jagged creation: {np.mean(time_jagged_creation):.4f} s +- {np.std(time_jagged_creation):.4f} s')
print(f'jagged loop: {np.mean(time_jagged_loop):.4f} s +- {np.std(time_jagged_loop):.4f} s')
```

Output:
```
numpy creation: 0.0694 s +- 0.0004 s
numpy array + for loop: 0.0186 s +- 0.0001 s
numpy comprehension: 0.0090 s +- 0.0001 s
numpy comprehension + array: 0.0125 s +- 0.0001 s
numpy vectorized: 0.0084 s +- 0.0002 s
jagged creation: 0.0011 s +- 0.0001 s
jagged loop: 0.0846 s +- 0.0004 s
jagged comprehension: 0.0698 s +- 0.0004 s
jagged comprehension + array: 0.0730 s +- 0.0004 s
```

I didn't expect such a gap between list comprehension and `for` loop, I need to read more about it.

To come back to the initial concern: do you think that there is still a need to integrate the convertion from a connectivity array with legacy format to a non-uniform array in pyvista ?

**Comment by darikg** [comment]
Interesting idea. Although since storing offsets and connectivity arrays separately is exactly what `CellArray` does (e.g. could presumably add `__getitem__` to `CellArray` directly?) I guess `JaggedArray` would be a more friendly/pythonic interface?

p.s. I just now noticed that `vtkCellArray` has [Import/ExportLegacyFormat methods](https://vtk.org/doc/nightly/html/classvtkCellArray.html#aaf677afcb4ac511927a57b55aab9d083), I wonder if they're of similar speed to the cython stuff

**Reply by Louis-Pujol** [reply]
I think you're right on the fact that `JaggedArray` can be just `CellArray` + `__getitem__`, everything we need to return view on a specific cell is already in `CellArray`, didn't think about it when reading https://github.com/pyvista/pyvista/issues/2330#issuecomment-1066168815.

For the `ImportLegacyFormat` stuff, it seems to me that you just find the optimal solution ! You can find [here](https://github.com/Louis-Pujol/cython_ncells_from_cells/blob/main/benchmark_with_legacy.py) a script to illustrate the difference of running times with varying number of cells (up to 100M), and appreciate how faster `ImportLegacyFormat` is:

<img src="https://github.com/pyvista/pyvista/assets/52413616/fe5de8b6-a543-40a2-b864-9dbfdf1d383a" alt="image" width="600" height="auto">

Seems that we can forget about `n_cells_from_cells`, as the combination of `ImportLegacyFormat` + `GetNumberOfCells` give the same result faster.

**Reply by darikg** [reply]
Very cool. Seems like we could drop `n_cells` (and PolyData's `n_faces`, `n_strips` etc.) entirely. I'll work on a PR introducing this (unless you'd prefer to)

**Reply by Louis-Pujol** [reply]
Nice, this will be a great improvement ! That's fine for me if you go ahead with it. If you need any help, let me know
